crates/cpuid-utils/src/lib.rs

```diff
@@ -23,55 +63,475 @@ pub mod bits;
 #[cfg(feature = "instance-spec")]
 mod instance_spec;
 
-#[cfg(feature = "instance-spec")]
-pub use instance_spec::*;
+type CpuidSubleafMap = BTreeMap<u32, CpuidValues>;
+type CpuidMapInsertResult = Result<Option<CpuidValues>, SubleafInsertConflict>;
+
+/// Denotes the presence or absence of subleaves for a given CPUID leaf.
+#[derive(Clone, Debug, PartialEq, Eq)]
+enum Subleaves {
+    /// This leaf is configured to have subleaves, whose values are stored in
+    /// the inner map.
+    Present(CpuidSubleafMap),
 
-/// A map from CPUID leaves to CPUID values.
+    /// This leaf does not have subleaves.
+    Absent(CpuidValues),
+}
+
+/// [`CpuidMap`]'s insert functions return this error if a request to insert a
+/// new value would produce a leaf that has both per-subleaf values and a
+/// no-subleaf value.
+#[derive(Debug, Error)]
+pub enum SubleafInsertConflict {
+    #[error("leaf {0:x} has entries with subleaves")]
+    SubleavesAlreadyPresent(u32),
+
+    #[error("leaf {0:x} has an entry with no subleaves")]
+    SubleavesAlreadyAbsent(u32),
+}
+
+/// A mapping from CPUID leaf/subleaf pairs to CPUID return values. This struct
+/// allows each registered leaf either to have or not to have subleaf values,
+/// but not both at once.
+///
+/// Some CPUID leaves completely ignore the subleaf value passed in ecx. Others
+/// pay attention to it and have their own per-leaf semantics that govern what
+/// happens if the supplied ecx value is out of the function's expected range.
+/// Tracking CPUID values in a simple `BTreeMap` from [`CpuidIdent`] to
+/// [`CpuidValues`] allows a single leaf to use both options at once, because
+/// `CpuidIdent { leaf: x, subleaf: None }` is not the same as `CpuidIdent {
+/// leaf: x, subleaf: Some(y) }`.
+///
+/// To avoid this kind of semantic confusion, this type's `insert` method
+/// returns a `Result` that indicates whether inserting the requested
+/// leaf/subleaf identifier would produce a semantic conflict between a
+/// subleaf-bearing and subleaf-free entry.
 #[derive(Clone, Debug, Default)]
-pub struct CpuidMap(pub BTreeMap<CpuidIdent, CpuidValues>);
+pub struct CpuidMap(BTreeMap<u32, Subleaves>);
```

Comment:
The implementation of this type is lovely --- the code is very straightforward and I didn't really have any meaningful concerns or nitpicks. Very nice!